runtime/lua/vim/lsp/callbacks.lua

```diff
@@ -11,13 +12,50 @@ local function err_message(...)
   api.nvim_command("redraw")
 end
 
+M['workspace/executeCommand'] = function(err, _)
+  if err then
+    error("Could not execute code action: "..err.message)
+  end
+end
+
+M['textDocument/codeAction'] = function(_, _, actions)
+  if vim.tbl_isempty(actions) then
+    print("No code actions available")
+    return
+  end
+
+  local option_strings = {"Code Actions:"}
+  for i, action in ipairs(actions) do
+    local title = action.title:gsub('\r\n', '\\r\\n')
+    title = title:gsub('\n', '\\n')
+    table.insert(option_strings, string.format("%d. %s", i, title))
+  end
+
+  local choice = vim.fn.inputlist(option_strings)
+  if choice < 1 or choice > #actions then
+    return
+  end
+  local action_chosen = actions[choice]
+  -- textDocument/codeAction can return either Command[] or CodeAction[].
+  -- We handle both in the same way by extracting the command
+  local command = action_chosen
+  if type(action_chosen.command) == table then
```

Comment:
For [eclipse.jdt.ls](https://github.com/eclipse/eclipse.jdt.ls) I had to add some special handling to get it working:


```lua
local action = action_chosen
if type(action_chosen.command) == table then
  action = action_chosen.command
end
if action.command == 'java.apply.workspaceEdit' then
  for _, argument in ipairs(action.arguments) do
    util.apply_workspace_edit(argument)
  end
else
  buf.execute_command(action)
end
```

A dump of the `action_chosen` looked like this:

```
["arguments"] = { 
  [1] = { 
    ["changes"] = { 
        ["file:///path/to/the/file"] = { 
          [1] = { 
            ["range"] = { 
              ["end"] = { 
                ["line"] = 24,
                ["character"] = 0,
              },
              ["start"] = { 
                ["line"] = 0,
                ["character"] = 0,
              },
            },
            ["newText"] = lots of text...,
          },
        },
      },
    },
  },
  ["title"] = Import 'Maps' (...),
  ["command"] = java.apply.workspaceEdit,
}
```

Reply:
That is more something for in https://github.com/neovim/nvim-lsp